---
title: "The DMRscaler user's guide knitrtest"
author: "Leroy Bondhus"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{The DMRscaler user's guide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


## Summary
dmrscaler is a method designed to identify features of differential methylation between case and control conditions across a wide range of genomic scale. In the first part of the vignette we step through downloading a publicly available dataset, running the statistical test for difference at individual CG site and a permutation test for estimating enrichment of differentially methylated CGs within a window. After this we run dmrscaler to call DMRs at each layer of window sizes generating a list of results. We then look at the data to visualize how dmrscaler has called and represents these DMR features.       


```{r}
load("temp_post_dmrscaler_pre_figure_gen.RData")
```


## install dmrscaler 
To install pull the most recent version from github and place in working directory 
```{r, echo=FALSE}
library("devtools")
library(roxygen2)
document()
install("../dmrscaler", quick=T)
library(doParallel)
library(rlang)
library("dplyr")
library(foreach)

```


## Choose an example region to look at
```{r}
example_region <- data.frame(chr="chr7",start=155000000, stop=159000000, stringsAsFactors = FALSE)
```


```{r, fig.width = 4, fig.height=4}
library(circlize)
library(HilbertCurve)

col_fun = colorRamp2(c(0,fdrscaler*0.99,fdrscaler,max(locs$scoring_values)), c("grey30", "grey60", "red", "red"))
level = 6
## 4^level - 1 = # segments

hc_points <- locs[which(locs$chr== example_region$chr ),]
hc_max <- nrow(hc_points)
hc_col <- col_fun(hc_points$scoring_values)
hc_size <- hc_points$scoring_values / max(hc_points$scoring_values)

#  hc_size <- hc_points$scoring_values / fdrscaler
hc <- HilbertCurve(s=1,e=hc_max, level = level, reference = F, title = example_region$chr)
hc_points(hc, x1 = 1:nrow(hc_points), np = NULL, pch=15, size = unit(hc_size*2, "mm"),gp = gpar(col = hc_col, fill = hc_col))
hc_polygon(hc, x1 = min(which(hc_points$pos >= example_region$start)), x2 = max(which(hc_points$pos <= example_region$stop)) )

### now looking only at the specified region
level = 4
hc_points <- locs[which(locs$chr== example_region$chr & locs$pos >= example_region$start & locs$pos <= example_region$stop ),]
hc_max <- nrow(hc_points)
hc_col <- col_fun(hc_points$scoring_values)
hc_size <- hc_points$scoring_values / max(hc_points$scoring_values)

#  hc_size <- hc_points$scoring_values / fdrscaler
hc <- HilbertCurve(s=1,e=hc_max, level = level, reference = F, title = example_region$chr)
hc_points(hc, x1 = 1:nrow(hc_points), np = NULL, pch=15, size = unit(hc_size*4, "mm"),gp = gpar(col = hc_col, fill = hc_col))



```


## Explore some features of the data
dmrscaler defines differential methylation features iteratively, expanding the size of the window for aggregating on at each step
this procedure allows a hierarchical structure to describe a region enriched in differntially methylated CpGs. We look at our example region here
```{r, fig.height=6, fig.width = 7}
library(GenomeInfoDb)
library(IRanges)
library(networkD3)

dmr_tree <- example_generate_dmr_tree(dmrscaler_result = dmrscaler_result, layer=5, chr=example_region$chr, start = example_region$start, stop = example_region$stop)
diagonalNetwork(List = dmr_tree, fontSize = 12, fontFamily = "bold", nodeStroke = "black", linkColour = "black", opacity = 1)

```

Next we can look at the region using the Gviz package
```{r, fig.width = 7, fig.height=5}
library(Gviz)


## organize data used in tracks
group <- character(length=ncol(B)) 
group[cases] <- "case"
group[controls] <- "control"
delta_mean_B <- rowMeans(B[,cases]) - rowMeans(B[,controls]) 
which <- which(locs$chr==example_region$chr & locs$pos >= example_region$start & locs$pos <= example_region$stop)
Bsub <- B[which, ]
gr <- GRanges(seqnames = example_region$chr, ranges = IRanges(start = locs[which,"pos"], width = 1), mcols = Bsub )

## set up ideogram track
itrack<-IdeogramTrack(genome="hg19", chromosome = example_region$chr)
## set up genome axis track
gtrack<-GenomeAxisTrack()
## set up significance track
sig_track <- DataTrack(range=gr, data = locs$scoring_values[which], type=c("p"))
## set up beta value track
beta_track <- DataTrack(range = gr, data = t(Bsub), groups=group,name="Beta", type=c("a","g","confint"))
## set up diff beta track
diff_beta_track <- DataTrack(range = gr, data = delta_mean_B[which], type=c("a","g"))

## set up gene model track ### 
library(biomaRt)
genesub <- example_gene_anno_df(chr=example_region$chr, start=example_region$start, stop=example_region$stop)
if(!all(is.na(genesub))){
  grtrack <- GeneRegionTrack(genesub, chromosome = example_region$chr, start = example_region$start, stop=example_region$stop, transcriptAnnotation="symbol" ,collapseTranscripts = "meta")
}else{grtrack<-GeneRegionTrack()}

## set up dmrscaler results layer tracks
layer_track <- list()
for(j in 1:length(dmrscaler_result)){
  layer_track[[j]] <- AnnotationTrack(start = dmrscaler_result[[j]]$start_pos, 
                                      end =  dmrscaler_result[[j]]$stop_pos,
                                      chromosome = dmrscaler_result[[j]]$chr,
                                      strand = "*", genome = "hg19", name=paste("layer", j, sep = "_"))
  displayPars(layer_track[[j]]) <- list(cex.legend=0.7, cex.axis=0.7, cex.title=0.7, rotation.title=0, stackHeight = 1, shape="box")
}


plotTracks(c(list(itrack, gtrack, sig_track, beta_track, diff_beta_track, grtrack), layer_track
                ), from = example_region$start-1, to=example_region$stop+1 ) 
```




